### PR TITLE
[ACR] Add network rule set properties, empty resource group validation

### PR DESCRIPTION
**Pull request by djyou**
Partially fixes https://github.com/Azure/azure-cli/issues/7256

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.

### CODE REVIEW

**Comment by djyou** [comment]
@sergey-shandar Could you please review this PR? We use the same model as [storage](https://github.com/Azure/azure-rest-api-specs/blob/b0478bcd1d3e010028ce5e74c24d2f66f898b21e/specification/storage/resource-manager/Microsoft.Storage/stable/2018-07-01/storage.json#L1106), but only a subset of properties at this point.
/cc @sajayantony

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json [hunk 1]
```diff
@@ -788,6 +783,11 @@
             "name": "ProvisioningState",
             "modelAsString": true
           }
+        },
```
Comment (sergey-shandar) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json]: The order of fields can be important in some languages. For example, C#/Javas constructors could be generated according to the order in the swagger file.

**Review by sergey-shandar — CHANGES_REQUESTED** [review]

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json (hunk 1 above)
Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json]: I reverted this change. SwashBuckle generated with a different order this time.

**Review by sergey-shandar — APPROVED** [review]

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json [hunk 2]
```diff
@@ -1497,6 +1497,10 @@
         "storageAccount": {
           "$ref": "#/definitions/StorageAccountProperties",
           "description": "The properties of the storage account for the container registry. Only applicable to Classic SKU."
+        },
```
Comment (KrisBash) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: What is the service behavior if a client performs a PUT with this API version and specifies networkRuleset, and another client GETs & PUTs the resource with a lower API version? Is the downlevel PUT blocked, or do you ignore networkRuleset for the update?

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json [hunk 3]
```diff
@@ -1535,6 +1539,48 @@
         }
       }
     },
+    "NetworkRuleSet": {
```
Comment (KrisBash) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: What is the reason that networkRuleset is only defined in this version and not the 2018-09-01 version? Is it now obsolete?

Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: `2018-09-01 ` doesn't have this resource type, `2017-10-01` is the latest version that has `registry`.

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json (hunk 2 above)
Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Lower versions ignore `networkRuleset` in PUT/PATCH, also won't return it in GET.

Comment (ravbhatnagar) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Seems reasonable to me. @KrisBash - signing off on your behalf.

Comment (ravbhatnagar) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Apologies, did not check thouroughly. @djyou  - seems like this new property is added in the same api-version. If your service were doing the right PUT behavior (replace semantics) then this would be a breaking change. I would recommend doing this in a new API-version.

Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Suppose I created a registry using this api-version `2010-10-01` with the new property `networkRuleSet`, I'd have the following json (other irrelevant properties ignored):
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Now if I do a PUT using lower api-versions, since they don't support the property `networkRuleSet`, it will be ignored (the same as whatever extra properties passed in). So:

PUT request:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    },
    "otherPropertyIgnored": "value"
}
```

PUT/GET response:
```
{
    "name": "myRegistry"
}
```

However, GET using this api-version `2017-10-01` will return the full json:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

@ravbhatnagar Could you clarify which part is a breaking change?

Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Within the same api-version `2017-10-01`, if I created a registry with the new property `networkRuleSet`:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Now if I do a PUT using old clients on the same api-version `2017-10-01`, since they don't have the new property `networkRuleSet`, it's essentially a `null` value:

PUT request:
```
{
    "name": "myRegistry",
    "networkRuleSet": null
}
```

On the server side, since the resource `myRegistry` already exists, it enters the update path, which replaces whatever values provided in the request, while ignoring unprovided ones.

PUT response:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Basically, this new property behaves in a same way as existing properties, such as `adminUserEnabled`, if it has a value, and a second PUT doesn't provide a value, we don't clear it's value.

Comment (ravbhatnagar) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Yes, this is exactly what I meant. Currently your RP does not implement PUT in the way it should have been where properties missing in the request are wiped out from the resource on the server. So in essence, following the "replace" semantics for PUT. Since your RP is not following the right model, it is currently not a breaking change. But for most services (specially the newer ones) in Azure, such changes go in a new api-version and PUT is a replace.

Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Thank you @ravbhatnagar We had an offline discussion within the team, since historically this resource didn't follow the replace semantic and introducing the right pattern itself is a breaking change (due to the existing properties `adminUserEnabled`, `storageAccountProperties`, etc.), do you think we can have this merged in for this api-verison, and we will introduce a new api-version that follows the right pattern? /cc @mnltejaswini @sajayantony

Comment (mnltejaswini) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: @ravbhatnagar We queued up an item to redesign the resource APIS with a newer api version and  other new features. As part of it, we can update the APIs to implement the semantics correctly.

Comment (ravbhatnagar) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Sounds good. Signing off from ARM side. opened an issue to track it.

**Comment by ravbhatnagar** [comment]
Just one issue tracking it in github. Signing off.

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json (hunk 2 above)
Comment (djyou) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json]: Thank you!

**Comment by djyou** [comment]
@sergey-shandar Ready to merge, thanks!

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go
